File: core/utils/common_helpers.js
const fs = require('fs').promises;
const path = require('path');
const jwt = require('jsonwebtoken');
const _ = require('lodash');
const url = require('url');

/**
 * A collection of utility functions for API testing.
 */
class ApiTestUtils {
  /**
   * Generate a random string of specified length.
   * 
   * @param {number} [length=10] - Length of the string to generate
   * @returns {string} A random string
   */
  static generateRandomString(length = 10) {
    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    let result = '';
    for (let i = 0; i < length; i++) {
      result += chars.charAt(Math.floor(Math.random() * chars.length));
    }
    return result;
  }

  /**
   * Generate a random email address.
   * 
   * @returns {string} A random email address
   */
  static generateRandomEmail() {
    const username = this.generateRandomString(8);
    const domain = this.generateRandomString(6);
    return `${username}@${domain}.com`;
  }

  /**
   * Generate a random phone number.
   * 
   * @returns {string} A random phone number
   */
  static generateRandomPhone() {
    const areaCode = Math.floor(Math.random() * 800) + 200;
    const prefix = Math.floor(Math.random() * 900) + 100;
    const lineNumber = Math.floor(Math.random() * 9000) + 1000;
    return `+1${areaCode}${prefix}${lineNumber}`;
  }

  /**
   * Generate a UUID.
   * 
   * @returns {string} A UUID string
   */
  static generateUuid() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
      const r = Math.random() * 16 | 0, v = c === 'x' ? r : (r & 0x3 | 0x8);
      return v.toString(16);
    });
  }

  /**
   * Get the current UNIX timestamp.
   * 
   * @returns {number} Current UNIX timestamp in seconds
   */
  static getTimestamp() {
    return Math.floor(Date.now() / 1000);
  }

  /**
   * Get the current timestamp in ISO 8601 format.
   * 
   * @returns {string} Current timestamp in ISO 8601 format
   */
  static getIsoTimestamp() {
    return new Date().toISOString();
  }

  /**
   * Load a JSON file.
   * 
   * @param {string} filePath - Path to the JSON file
   * @returns {Promise<Object>} The loaded JSON
   * @throws {Error} If the file does not exist or is not valid JSON
   */
  static async loadJsonFile(filePath) {
    try {
      const data = await fs.readFile(path.resolve(filePath), 'utf8');
      return JSON.parse(data);
    } catch (error) {
      console.error(`Error loading JSON file: ${error.message}`);
      throw error;
    }
  }

  /**
   * Save data to a JSON file.
   * 
   * @param {*} data - Data to save
   * @param {string} filePath - Path to the JSON file
   * @returns {Promise<void>}
   * @throws {Error} If the file cannot be written
   */
  static async saveJsonFile(data, filePath) {
    try {
      const dirPath = path.dirname(filePath);
      await fs.mkdir(dirPath, { recursive: true });
      await fs.writeFile(path.resolve(filePath), JSON.stringify(data, null, 2));
    } catch (error) {
      console.error(`Error saving JSON file: ${error.message}`);
      throw error;
    }
  }

  /**
   * Parse a JWT token.
   * 
   * @param {string} token - JWT token
   * @returns {Object} The decoded JWT payload
   * @throws {Error} If the token is invalid
   */
  static parseJwt(token) {
    try {
      // Decode without verification (for testing purposes)
      return jwt.decode(token);
    } catch (error) {
      console.error(`Invalid JWT token: ${error.message}`);
      throw error;
    }
  }

  /**
   * Wait for a condition to be true.
   * 
   * @param {Function} conditionFunc - Function that returns a promise resolving to true when the condition is met
   * @param {number} [timeout=30000] - Maximum time to wait in milliseconds
   * @param {number} [interval=1000] - Interval between checks in milliseconds
   * @returns {Promise<boolean>} True if the condition was met, false if the timeout was reached
   */
  static async waitForCondition(conditionFunc, timeout = 30000, interval = 1000) {
    const startTime = Date.now();
    while (Date.now() - startTime < timeout) {
      if (await conditionFunc()) {
        return true;
      }
      await new Promise(resolve => setTimeout(resolve, interval));
    }
    return false;
  }

  /**
   * Retry a function on failure.
   * 
   * @param {Function} func - Function to retry that returns a promise
   * @param {Object} [options] - Retry options
   * @param {number} [options.retries=3] - Number of retries
   * @param {number} [options.delay=1000] - Initial delay between retries in milliseconds
   * @param {number} [options.backoff=2] - Backoff multiplier for the delay
   * @returns {Promise<*>} The result of the function
   * @throws {Error} The last error if all retries fail
   */
  static async retryOnFailure(func, { retries = 3, delay = 1000, backoff = 2 } = {}) {
    let retryCount = 0;
    let currentDelay = delay;
    
    while (true) {
      try {
        return await func();
      } catch (error) {
        retryCount++;
        if (retryCount > retries) {
          console.error(`All ${retries} retries failed. Last error: ${error.message}`);
          throw error;
        }
        
        console.warn(`Retry ${retryCount}/${retries} after error: ${error.message}`);
        await new Promise(resolve => setTimeout(resolve, currentDelay));
        currentDelay *= backoff;
      }
    }
  }

  /**
   * Compare two JSON objects for equality, optionally ignoring certain keys.
   * 
   * @param {Object} obj1 - First JSON object
   * @param {Object} obj2 - Second JSON object
   * @param {Array<string>} [ignoreKeys=[]] - Keys to ignore in the comparison
   * @returns {boolean} True if the objects are equal, false otherwise
   */
  static compareJsonObjects(obj1, obj2, ignoreKeys = []) {
    // Create deep clones to avoid modifying the originals
    const obj1Copy = _.cloneDeep(obj1);
    const obj2Copy = _.cloneDeep(obj2);
    
    // Remove ignored keys
    ignoreKeys.forEach(key => {
      _.unset(obj1Copy, key);
      _.unset(obj2Copy, key);
    });
    
    return _.isEqual(obj1Copy, obj2Copy);
  }

  /**
   * Extract a value from a nested object using a path.
   * 
   * @param {Object|Array} obj - Object to extract from
   * @param {string} path - Path to the value (e.g., 'data.items[0].id')
   * @returns {*} The extracted value
   * @throws {Error} If the path is invalid
   */
  static extractNestedValue(obj, path) {
    try {
      return _.get(obj, path);
    } catch (error) {
      throw new Error(`Failed to extract value at path '${path}': ${error.message}`);
    }
  }

  /**
   * Mask sensitive data in an object.
   * 
   * @param {*} data - Data to mask
   * @param {Array<string>} sensitiveKeys - Keys to mask
   * @param {string} [mask='******'] - Mask to use
   * @returns {*} The masked data
   */
  static maskSensitiveData(data, sensitiveKeys, mask = '******') {
    if (!data) return data;
    
    if (Array.isArray(data)) {
      return data.map(item => this.maskSensitiveData(item, sensitiveKeys, mask));
    }
    
    if (typeof data === 'object') {
      const result = {};
      
      for (const [key, value] of Object.entries(data)) {
        if (sensitiveKeys.includes(key)) {
          result[key] = mask;
        } else if (typeof value === 'object' && value !== null) {
          result[key] = this.maskSensitiveData(value, sensitiveKeys, mask);
        } else {
          result[key] = value;
        }
      }
      
      return result;
    }
    
    return data;
  }

  /**
   * Extract the base URL from a request URL.
   * 
   * @param {string} requestUrl - Request URL
   * @returns {string} The base URL
   */
  static getBaseUrlFromRequest(requestUrl) {
    const parsedUrl = new URL(requestUrl);
    return `${parsedUrl.protocol}//${parsedUrl.host}`;
  }

  /**
   * Merge two objects deeply.
   * 
   * @param {Object} obj1 - First object
   * @param {Object} obj2 - Second object
   * @param {boolean} [overwrite=true] - Whether to overwrite values in obj1 with values from obj2
   * @returns {Object} The merged object
   */
  static mergeObjects(obj1, obj2, overwrite = true) {
    if (overwrite) {
      return _.merge({}, obj1, obj2);
    } else {
      return _.merge({}, obj2, obj1);
    }
  }

  /**
   * Sleep for a specified duration.
   * 
   * @param {number} ms - Duration in milliseconds
   * @returns {Promise<void>}
   */
  static sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  /**
   * Format a date according to the specified format.
   * 
   * @param {Date|string|number} date - Date to format
   * @param {string} format - Format string (e.g., 'YYYY-MM-DD')
   * @returns {string} The formatted date
   */
  static formatDate(date, format) {
    const d = new Date(date);
    
    const formatMap = {
      YYYY: d.getFullYear(),
      MM: String(d.getMonth() + 1).padStart(2, '0'),
      DD: String(d.getDate()).padStart(2, '0'),
      HH: String(d.getHours()).padStart(2, '0'),
      mm: String(d.getMinutes()).padStart(2, '0'),
      ss: String(d.getSeconds()).padStart(2, '0')
    };
    
    let result = format;
    for (const [key, value] of Object.entries(formatMap)) {
      result = result.replace(key, value);
    }
    
    return result;
  }
}

module.exports = ApiTestUtils;

File: core/utils/common_helpers.py
import time
import json
import random
import string
import logging
import datetime
import uuid
from typing import Dict, Any, List, Union, Optional
import requests
import jwt

logger = logging.getLogger(__name__)

class ApiTestUtils:
    """
    A collection of utility functions for API testing.
    """
    
    @staticmethod
    def generate_random_string(length: int = 10) -> str:
        """
        Generate a random string of specified length.
        
        Args:
            length: Length of the string to generate
            
        Returns:
            A random string
        """
        return ''.join(random.choices(string.ascii_letters + string.digits, k=length))
    
    @staticmethod
    def generate_random_email() -> str:
        """
        Generate a random email address.
        
        Returns:
            A random email address
        """
        username = ApiTestUtils.generate_random_string(8)
        domain = ApiTestUtils.generate_random_string(6)
        return f"{username}@{domain}.com"
    
    @staticmethod
    def generate_random_phone() -> str:
        """
        Generate a random phone number.
        
        Returns:
            A random phone number
        """
        return f"+1{random.randint(2, 9)}{random.randint(0, 9)}{''.join(str(random.randint(0, 9)) for _ in range(8))}"
    
    @staticmethod
    def generate_uuid() -> str:
        """
        Generate a UUID.
        
        Returns:
            A UUID string
        """
        return str(uuid.uuid4())
    
    @staticmethod
    def get_timestamp() -> int:
        """
        Get the current UNIX timestamp.
        
        Returns:
            Current UNIX timestamp in seconds
        """
        return int(time.time())
    
    @staticmethod
    def get_iso_timestamp() -> str:
        """
        Get the current timestamp in ISO 8601 format.
        
        Returns:
            Current timestamp in ISO 8601 format
        """
        return datetime.datetime.now().isoformat()
    
    @staticmethod
    def load_json_file(file_path: str) -> Dict[str, Any]:
        """
        Load a JSON file.
        
        Args:
            file_path: Path to the JSON file
            
        Returns:
            The loaded JSON as a dictionary
            
        Raises:
            FileNotFoundError: If the file does not exist
            json.JSONDecodeError: If the file is not valid JSON
        """
        try:
            with open(file_path, 'r') as f:
                return json.load(f)
        except FileNotFoundError:
            logger.error(f"File not found: {file_path}")
            raise
        except json.JSONDecodeError:
            logger.error(f"Invalid JSON in file: {file_path}")
            raise
    
    @staticmethod
    def save_json_file(data: Any, file_path: str) -> None:
        """
        Save data to a JSON file.
        
        Args:
            data: Data to save
            file_path: Path to the JSON file
            
        Raises:
            IOError: If the file cannot be written
        """
        try:
            with open(file_path, 'w') as f:
                json.dump(data, f, indent=2)
        except IOError:
            logger.error(f"Error writing to file: {file_path}")
            raise
    
    @staticmethod
    def parse_jwt(token: str) -> Dict[str, Any]:
        """
        Parse a JWT token.
        
        Args:
            token: JWT token
            
        Returns:
            The decoded JWT payload
            
        Raises:
            jwt.InvalidTokenError: If the token is invalid
        """
        try:
            # Decode without verification (for testing purposes)
            return jwt.decode(token, options={"verify_signature": False})
        except jwt.InvalidTokenError:
            logger.error("Invalid JWT token")
            raise
    
    @staticmethod
    def wait_for_condition(condition_func, timeout: int = 30, interval: int = 1) -> bool:
        """
        Wait for a condition to be true.
        
        Args:
            condition_func: Function that returns True when the condition is met
            timeout: Maximum time to wait in seconds
            interval: Interval between checks in seconds
            
        Returns:
            True if the condition was met, False if the timeout was reached
        """
        start_time = time.time()
        while time.time() - start_time < timeout:
            if condition_func():
                return True
            time.sleep(interval)
        return False
    
    @staticmethod
    def retry_on_failure(func, retries: int = 3, delay: int = 1, backoff: int = 2, 
                        expected_exceptions: tuple = (Exception,)):
        """
        Retry a function on failure.
        
        Args:
            func: Function to retry
            retries: Number of retries
            delay: Initial delay between retries in seconds
            backoff: Backoff multiplier for the delay
            expected_exceptions: Exceptions to catch and retry on
            
        Returns:
            The result of the function
            
        Raises:
            The last exception if all retries fail
        """
        retry_count = 0
        current_delay = delay
        
        while True:
            try:
                return func()
            except expected_exceptions as e:
                retry_count += 1
                if retry_count > retries:
                    logger.error(f"All {retries} retries failed. Last error: {str(e)}")
                    raise
                
                logger.warning(f"Retry {retry_count}/{retries} after error: {str(e)}")
                time.sleep(current_delay)
                current_delay *= backoff
    
    @staticmethod
    def compare_json_objects(obj1: Dict[str, Any], obj2: Dict[str, Any], 
                          ignore_keys: Optional[List[str]] = None) -> bool:
        """
        Compare two JSON objects for equality, optionally ignoring certain keys.
        
        Args:
            obj1: First JSON object
            obj2: Second JSON object
            ignore_keys: Keys to ignore in the comparison
            
        Returns:
            True if the objects are equal, False otherwise
        """
        if ignore_keys is None:
            ignore_keys = []
            
        # Create copies to avoid modifying the originals
        obj1_copy = obj1.copy()
        obj2_copy = obj2.copy()
        
        # Remove ignored keys
        for key in ignore_keys:
            obj1_copy.pop(key, None)
            obj2_copy.pop(key, None)
            
        return obj1_copy == obj2_copy
    
    @staticmethod
    def extract_nested_value(obj: Union[Dict[str, Any], List[Any]], path: str) -> Any:
        """
        Extract a value from a nested object using a path.
        
        Args:
            obj: Object to extract from
            path: Path to the value (e.g., 'data.items[0].id')
            
        Returns:
            The extracted value
            
        Raises:
            ValueError: If the path is invalid
        """
        parts = path.replace('[', '.').replace(']', '').split('.')
        current = obj
        
        for part in parts:
            if not part:  # Skip empty parts
                continue
                
            if isinstance(current, dict):
                if part not in current:
                    raise ValueError(f"Key '{part}' not found in path '{path}'")
                current = current[part]
            elif isinstance(current, list):
                try:
                    index = int(part)
                    if index >= len(current):
                        raise ValueError(f"Index {index} out of range in path '{path}'")
                    current = current[index]
                except ValueError:
                    raise ValueError(f"Invalid list index '{part}' in path '{path}'")
            else:
                raise ValueError(f"Cannot navigate further from {type(current)} in path '{path}'")
                
        return current
    
    @staticmethod
    def mask_sensitive_data(data: Union[Dict[str, Any], List[Any], str], 
                          sensitive_keys: List[str], mask: str = '******') -> Union[Dict[str, Any], List[Any], str]:
        """
        Mask sensitive data in an object.
        
        Args:
            data: Data to mask
            sensitive_keys: Keys to mask
            mask: Mask to use
            
        Returns:
            The masked data
        """
        if isinstance(data, dict):
            masked_data = {}
            for key, value in data.items():
                if key in sensitive_keys:
                    masked_data[key] = mask
                else:
                    masked_data[key] = ApiTestUtils.mask_sensitive_data(value, sensitive_keys, mask)
            return masked_data
        elif isinstance(data, list):
            return [ApiTestUtils.mask_sensitive_data(item, sensitive_keys, mask) for item in data]
        else:
            return data
    
    @staticmethod
    def get_base_url_from_request(request_url: str) -> str:
        """
        Extract the base URL from a request URL.
        
        Args:
            request_url: Request URL
            
        Returns:
            The base URL
        """
        parsed_url = requests.utils.urlparse(request_url)
        return f"{parsed_url.scheme}://{parsed_url.netloc}"
    
    @staticmethod
    def merge_dictionaries(dict1: Dict[str, Any], dict2: Dict[str, Any], 
                          overwrite: bool = True) -> Dict[str, Any]:
        """
        Merge two dictionaries.
        
        Args:
            dict1: First dictionary
            dict2: Second dictionary
            overwrite: Whether to overwrite values in dict1 with values from dict2
            
        Returns:
            The merged dictionary
        """
        result = dict1.copy()
        
        for key, value in dict2.items():
            if key in result and isinstance(result[key], dict) and isinstance(value, dict):
                result[key] = ApiTestUtils.merge_dictionaries(result[key], value, overwrite)
            elif key not in result or overwrite:
                result[key] = value
                
        return result
